### ISSUE TITLE
dp noise updates non-trainable layers

### ISSUE THREAD

**Issue by simra** [simra/FedNewsRec#12]
@wanglun1996 
We are adding noise to the embedding layer, which drives divergence.

However, in trying to fix this I found some inconsistencies that I don't understand.  Some trainable gradients are all zeros, while other trainable parameters aren't listed in named_parameters() and so I'm having difficulty extracting requires_grad.

This code attempts to initialize the set of trainable layers:
`requires_dp_noise = { layer: p.requires_grad for layer, p in model.named_parameters()}`

Later I try to measure sparseness to see what layers actually have gradients:
`sparseness = {layer: (requires_dp_noise.get(layer, "Unknown"), 
                                  float(torch.count_nonzero(update[layer]).cpu().item())/torch.numel(update[layer])) for layer in update}
`

If I dump sparseness, some are fully sparse but trainable, others are not sparse but have "unknown" trainable state ie they're missing from named_parameters():
`
{
   "doc_encoder.phase1.2.weight": [
      true,
      0.9999916666666666
   ],
   "doc_encoder.phase1.2.bias": [
      true,
      1.0
   ],
   "doc_encoder.attention.WQ.weight": [
      true,
      0.98786875
   ],
   "doc_encoder.attention.WK.weight": [
      true,
      0.98575
   ],
   "doc_encoder.attention.WV.weight": [
      true,
      0.786925
   ],
   "doc_encoder.phase2.2.dense.weight": [
      true,
      0.7115
   ],
   "doc_encoder.phase2.2.dense.bias": [
      true,
      0.8
   ],
   "doc_encoder.phase2.2.dense2.weight": [
      true,
      1.0
   ],
   "doc_encoder.phase2.2.dense2.bias": [
      true,
      0.0
   ],
   "user_encoder.attention2.WQ.weight": [
      true,
      0.17438125
   ],
   "user_encoder.attention2.WK.weight": [
      true,
      0.158275
   ],
   "user_encoder.attention2.WV.weight": [
      true,
      0.68285
   ],
   "user_encoder.pool2.dense.weight": [
      true,
      0.7598
   ],
   "user_encoder.pool2.dense.bias": [
      true,
      0.13
   ],
   "user_encoder.pool2.dense2.weight": [
      true,
      0.995
   ],
   "user_encoder.pool2.dense2.bias": [
      true,
      0.0
   ],
   "user_encoder.gru2.weight_ih_l0": [
      true,
      0.6575895833333333
   ],
   "user_encoder.gru2.weight_hh_l0": [
      true,
      0.9580083333333334
   ],
   "user_encoder.gru2.bias_ih_l0": [
      true,
      0.9991666666666666
   ],
   "user_encoder.gru2.bias_hh_l0": [
      true,
      0.9983333333333333
   ],
   "user_encoder.pool3.dense.weight": [
      true,
      0.9870875
   ],
   "user_encoder.pool3.dense.bias": [
      true,
      0.91
   ],
   "user_encoder.pool3.dense2.weight": [
      true,
      1.0
   ],
   "user_encoder.pool3.dense2.bias": [
      true,
      0.0
   ],
   "title_word_embedding_layer.weight": [
      false,
      0.0
   ],
   "click_td.module.phase1.2.weight": [
      "Unknown",
      0.9999916666666666
   ],
   "click_td.module.phase1.2.bias": [
      "Unknown",
      1.0
   ],
   "click_td.module.attention.WQ.weight": [
      "Unknown",
      0.98786875
   ],
   "click_td.module.attention.WK.weight": [
      "Unknown",
      0.98575
   ],
   "click_td.module.attention.WV.weight": [
      "Unknown",
      0.786925
   ],
   "click_td.module.phase2.2.dense.weight": [
      "Unknown",
      0.7115
   ],
   "click_td.module.phase2.2.dense.bias": [
      "Unknown",
      0.8
   ],
   "click_td.module.phase2.2.dense2.weight": [
      "Unknown",
      1.0
   ],
   "click_td.module.phase2.2.dense2.bias": [
      "Unknown",
      0.0
   ],
   "can_td.module.phase1.2.weight": [
      "Unknown",
      0.9999916666666666
   ],
   "can_td.module.phase1.2.bias": [
      "Unknown",
      1.0
   ],
   "can_td.module.attention.WQ.weight": [
      "Unknown",
      0.98786875
   ],
   "can_td.module.attention.WK.weight": [
      "Unknown",
      0.98575
   ],
   "can_td.module.attention.WV.weight": [
      "Unknown",
      0.786925
   ],
   "can_td.module.phase2.2.dense.weight": [
      "Unknown",
      0.7115
   ],
   "can_td.module.phase2.2.dense.bias": [
      "Unknown",
      0.8
   ],
   "can_td.module.phase2.2.dense2.weight": [
      "Unknown",
      1.0
   ],
   "can_td.module.phase2.2.dense2.bias": [
      "Unknown",
      0.0
   ]
}`
        
I think this is an issue with the TimeDistributed or DocEncoder module but it's not yet clear to me what the fix is.  I think the workaround for now is to assume a layer is trainable unless requires_grad is explicitly false.

**Comment by simra** [comment]
Ok, I see now that click_td and can_td are just wrappers for the same doc_encoder.  So in principle once we update the doc encoder I don't think we need to do anything with click_td and can_td.  I'm not sure how load_state_dict() is treating these- we might just be able to ignore them when computing running_average.
